l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol

```diff
@@ -0,0 +1,864 @@
+   * @dev Can be called by anyone once a round has passed its execution delay but is still within its lifetime.
+   *      The function tallies all votes cast during the round, identifies validators that reached the quorum threshold,
+   *      and executes slashing by deploying a SlashPayload contract and calling the Slasher.
+   *
+   * @param _round The round number to execute (must be ready for execution based on timing constraints)
+   * @param _committees Array of validator committees slashed for each epoch in the round being executed.
+   *                   Must contain exactly ROUND_SIZE_IN_EPOCHS committees. Only committees with slashed
+   *                   validators will have their commitments verified against on-chain data.
+   *
+   * Emits:
+   * - RoundExecuted: When the round execution completes, regardless of whether any slashing occurred
+   *
+   * Reverts with:
+   * - ConsensusSlashingProposer__RoundAlreadyExecuted: If the round has already been executed
+   * - ConsensusSlashingProposer__RoundNotComplete: If the round is not yet ready for execution or has expired
+   * - ConsensusSlashingProposer__InvalidCommitteeCommitment: If any committee commitment doesn't match on-chain data
+   * - ConsensusSlashingProposer__InvalidNumberOfCommittees: If the number of committees doesn't match
+   * ROUND_SIZE_IN_EPOCHS
+   */
+  function executeRound(SlashRound _round, address[][] calldata _committees) external {
+    // Get round data to check if already executed
+    SlashRound currentRound = getCurrentRound();
+    RoundData memory roundData = getRoundData(_round, currentRound);
+    require(!roundData.executed, Errors.ConsensusSlashingProposer__RoundAlreadyExecuted(_round));
+
+    // Ensure enough time has passed (execution delay) but not too much (lifetime)
+    require(isRoundReadyToExecute(_round, currentRound), Errors.ConsensusSlashingProposer__RoundNotComplete(_round));
+
+    // Get the slash actions by tallying votes and which committees have slashes
+    (SlashAction[] memory actions, bool[] memory committeesWithSlashes) = tally(roundData, _committees);
+
+    // Only verify committees that have slashed validators
+    for (uint256 i = 0; i < committeesWithSlashes.length; i++) {
+      if (!committeesWithSlashes[i]) {
+        continue;
+      }
+
+      // Check committee commitments against the stored on-chain data
+      bytes32 commitment = computeCommitteeCommitment(_committees[i]);
+      Epoch epochNumber = getEpochSlashed(_round, i);
+      require(
+        commitment == getCommitteeCommitment(epochNumber),
+        Errors.ConsensusSlashingProposer__InvalidCommitteeCommitment()
+      );
+    }
+
+    // Execute slashes if any were determined
+    if (actions.length > 0) {
+      // Deploy payload contract and execute slashes
+      IPayload slashPayload = deploySlashPayload(_round, actions);
+      SLASHER.slash(slashPayload);
```

Comment:
With tally and all that going on in here, do we have a clear idea about how many of the potential actors we can actually punish? 

Say that a cartel is to be slashed, but some other people did things at the same time such that it is questionable whether they were online or not. It could be in the interest of the cartel to start voting much more aggressively to slash everyone (including themselves) if they can move some of the "maybe" beyond the threshold and have the execution run out of gas.